effekt/shared/src/main/scala/effekt/typer/CapabilityScope.scala

```diff
@@ -24,7 +24,19 @@ case object GlobalCapabilityScope extends CapabilityScope {
 class BindSome(binder: source.Tree, capabilities: Map[symbols.InterfaceType, symbols.BlockParam],val parent: CapabilityScope) extends CapabilityScope {
   def copy: CapabilityScope = BindSome(binder, capabilities, parent.copy)
   def capabilityFor(tpe: symbols.InterfaceType)(using C: Context): symbols.BlockParam =
-    capabilities.getOrElse(tpe, parent.capabilityFor(tpe))
+    capabilities.get(tpe).orElse { // TODO(jiribenes, 2025-11-07): WIP, this is pretty horrible...
+      // If the requested type has unbound type parameters, try lexical resolution
+      unknowns(tpe).toList match {
+        case Nil => None
+        case _ => capabilities.collectFirst {
+          case (handlerTpe, cap) if handlerTpe.typeConstructor == tpe.typeConstructor =>
+            (tpe.args zip handlerTpe.args).foreach { case (inferred, concrete) =>
+              C.requireSubtype(inferred, concrete, ErrorContext.MergeTypes(inferred, concrete)) // TODO(jiribenes, 2025-11-07): better error message?
+            }
+            cap
+        }
+      }
+    }.getOrElse(parent.capabilityFor(tpe))
```

Comment:
This is likely imprecise!

Note the following where if you remove the `[A]` in `val a = do read[A]()`, you get "got A but expected B" on the use site (last line):
https://github.com/effekt-lang/effekt/blob/795b14e1c2bec5de6d44742b67c19db318bf537c/libraries/common/stream.effekt#L154-L163

Reply:
Also doesn't work here -- removing the `[Char]` annotations causes the good old "Effects need to be fully known, but effect Scan[A]'s type parameter(s) A could not be inferred." to trigger again.
Note that the `try` there in the snippet handles something else, but the inner `exponent` function is tricky: it does have the `Scan[A]` effect inferred, but it's "too far" for this method to discover it? (or there's some logic error I've made)

https://github.com/effekt-lang/effekt/blob/795b14e1c2bec5de6d44742b67c19db318bf537c/libraries/common/json.effekt#L94-L99

Reply:
The latter now at least has a better error message 🌸
<img width="931" height="167" alt="Screenshot 2025-11-11 at 21 45 22" src="https://github.com/user-attachments/assets/24c193a8-084b-472b-af8f-6cbb044a5e57" />